libc/syscalls.c

```diff
@@ -107,20 +107,20 @@ static long _syscall_nanosleep(long n, long x1, long x2)
 
     OE_UNUSED(n);
 
-    if (rem)
-        memset(rem, 0, sizeof(*rem));
-
-    if (!req)
-        goto done;
-
-    /* Convert timespec to milliseconds */
-    milliseconds += req->tv_sec * 1000UL;
-    milliseconds += req->tv_nsec / 1000000UL;
+    if (req)
+    {
+        /* Convert timespec to milliseconds */
+        milliseconds += req->tv_sec * 1000UL;
+        milliseconds += req->tv_nsec / 1000000UL;
 
-    /* Perform OCALL */
-    ret = oe_sleep_msec(milliseconds);
+        /* Perform OCALL */
+        ret = oe_sleep_msec(milliseconds);
+    }
 
-done:
+    /* Clear remaining time. Do this after reading from req because it may be
+     * that req == rem. */
+    if (rem)
```

Comment:
Shouldn't we check the `ret` here as well? Or we just ignore the `rem` regardless what the `ret` will be.

Reply:
+1.

I'm not sure this can be properly addressed by this PR though; the stack looks incorrectly implemented on the host side since we break the `nanosleep()` semantics by not marshalling the expected `rem` value back to the enclave, so the value is always all or nothing and not dependent on `EINTR`.